pandas 1.4.0: "ValueError: No objects to concatenate"

For pandas 1.4, problems occur for multi-period investment in `define_storage_unit_constraint` and `define_store_constraint`.

https://github.com/PyPSA/PyPSA/blob/70f1fa481819b0fab9228d997f5bac132798a582/pypsa/linopf.py#L507

https://github.com/PyPSA/PyPSA/blob/70f1fa481819b0fab9228d997f5bac132798a582/pypsa/linopf.py#L584

```
pypsa/components.py:660: in lopf
    return network_lopf_lowmem(self, **args)
pypsa/linopf.py:1200: in network_lopf
    fdp, problem_fn = prepare_lopf(n, snapshots, keep_files, skip_objective,
pypsa/linopf.py:905: in prepare_lopf
    define_storage_unit_constraints(n, snapshots)
pypsa/linopf.py:507: in define_storage_unit_constraints
    active[noncyclic_pp_i].groupby(level=0).transform(pd.Series.cumsum) == 1)
/home/fneum/miniconda3/envs/pypsa/lib/python3.9/site-packages/pandas/core/groupby/generic.py:1184: in transform
    return self._transform(
/home/fneum/miniconda3/envs/pypsa/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:1633: in _transform
    return self._transform_general(func, *args, **kwargs)
/home/fneum/miniconda3/envs/pypsa/lib/python3.9/site-packages/pandas/core/groupby/generic.py:1177: in _transform_general
    concatenated = concat(applied, axis=self.axis, verify_integrity=False)
/home/fneum/miniconda3/envs/pypsa/lib/python3.9/site-packages/pandas/util/_decorators.py:311: in wrapper
    return func(*args, **kwargs)
/home/fneum/miniconda3/envs/pypsa/lib/python3.9/site-packages/pandas/core/reshape/concat.py:346: in concat
    op = _Concatenator(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pandas.core.reshape.concat._Concatenator object at 0x7fc74da24370>, objs = [], axis = 0, join = 'outer', keys = None, levels = None, names = None, ignore_index = False, verify_integrity = False, copy = True
sort = False

    def __init__(
        self,
        objs: Iterable[NDFrame] | Mapping[Hashable, NDFrame],
        axis=0,
        join: str = "outer",
        keys=None,
        levels=None,
        names=None,
        ignore_index: bool = False,
        verify_integrity: bool = False,
        copy: bool = True,
        sort=False,
    ):
        if isinstance(objs, (ABCSeries, ABCDataFrame, str)):
            raise TypeError(
                "first argument must be an iterable of pandas "
                f'objects, you passed an object of type "{type(objs).__name__}"'
            )
    
        if join == "outer":
            self.intersect = False
        elif join == "inner":
            self.intersect = True
        else:  # pragma: no cover
            raise ValueError(
                "Only can inner (intersect) or outer (union) join the other axis"
            )
    
        if isinstance(objs, abc.Mapping):
            if keys is None:
                keys = list(objs.keys())
            objs = [objs[k] for k in keys]
        else:
            objs = list(objs)
    
        if len(objs) == 0:
>           raise ValueError("No objects to concatenate")
E           ValueError: No objects to concatenate

/home/fneum/miniconda3/envs/pypsa/lib/python3.9/site-packages/pandas/core/reshape/concat.py:403: ValueError
```

--- a/doc/release_notes.rst
+++ b/doc/release_notes.rst
@@ -25,6 +25,8 @@ Upcoming Release
 
 * Fix bug when multi-links are defined but the network has no links. 
 
+* Combatibility with pandas 1.4.
+
 
 PyPSA 0.18.1 (15th October 2021)
 ================================

--- a/pypsa/linopf.py
+++ b/pypsa/linopf.py
@@ -504,7 +504,7 @@ def define_storage_unit_constraints(n, sns):
 
         # set the initial enery at the beginning of each period
         first_active_snapshot_pp = (
-            active[noncyclic_pp_i].groupby(level=0).transform(pd.Series.cumsum) == 1)
+            active[noncyclic_pp_i].groupby(level=0).cumsum() == 1)
 
         lhs += masked_term(eff_stand[~first_active_snapshot_pp],
                            soc.shift()[~first_active_snapshot_pp],
@@ -581,7 +581,7 @@ def define_store_constraints(n, sns):
 
         # set the initial enery at the beginning of each period
         first_active_snapshot_pp = (
-            active[noncyclic_pp_i].groupby(level=0).transform(pd.Series.cumsum) == 1)
+            active[noncyclic_pp_i].groupby(level=0).cumsum() == 1)
 
         lhs += masked_term(eff_stand[~first_active_snapshot_pp],
                            e.shift()[~first_active_snapshot_pp],

